In [ ]:
from google.colab import drive
drive.mount("/content/gdrive")

Mounted at /content/gdrive


In [ ]:
import pandas as pd
pd.read_csv('/content/gdrive/My Drive/Datasets/StockData.csv')

UnicodeDecodeError: ignored

In [ ]:
df=pd.read_csv('/content/gdrive/My Drive/Datasets/StockData.csv', encoding = "ISO-8859-1")

In [ ]:
df.head()

Date  ...                         Top25
0  2000-01-03  ...            Recovering a title
1  2000-01-04  ...  Millennium bug fails to bite
2  2000-01-05  ...                  Useful links
3  2000-01-06  ...   Lessons of law's hard heart
4  2000-01-07  ...                    Gone aloft

[5 rows x 27 columns]

In [ ]:
df.head(1000)

Date  ...                                  Top25
0    2000-01-03  ...                     Recovering a title
1    2000-01-04  ...           Millennium bug fails to bite
2    2000-01-05  ...                           Useful links
3    2000-01-06  ...            Lessons of law's hard heart
4    2000-01-07  ...                             Gone aloft
..          ...  ...                                    ...
995  2004-02-19  ...  Mickey Mouse, and Sven-Goran Eriksson
996  2004-02-20  ...                 The Holocaust's shadow
997  2004-02-23  ...                  Interview: Tony Leung
998  2004-02-24  ...                   Burt is ITV chairman
999  2004-02-25  ...   Chinese students excel at GCSE level

[1000 rows x 27 columns]

In [ ]:
df.tail()

Date  ...                                              Top25
4096  2016-06-27  ...  Pakistani clerics declare transgender marriage...
4097  2016-06-28  ...  Mexico's security forces routinely use 'sexual...
4098  2016-06-29  ...  Tiny New Zealand town with 'too many jobs' lau...
4099  2016-06-30  ...  Law Enforcement Sources: Gun Used in Paris Ter...
4100  2016-07-01  ...  Ozone layer hole seems to be healing - US &amp...

[5 rows x 27 columns]

In [ ]:
#Training and testing Dataset

train = df[df['Date'] < '20150101']
test = df[df['Date'] > '20141231']

In [ ]:
test.head()

Date  ...                                              Top25
3723  2015-01-02  ...  Authorities Pull Plug on Russia's Last Politic...
3724  2015-01-05  ...                      China scraps rare earths caps
3725  2015-01-06  ...  Tests over, India set to make the iris of bigg...
3726  2015-01-07  ...   Ebola vaccine trials in Halifax so far a success
3727  2015-01-08  ...  Young mother let terrorists into Charlie Hebdo...

[5 rows x 27 columns]

In [ ]:
#Preprocessing
data=train.iloc[:,2:27]
data.replace("[^a-zA-Z]"," ",regex=True, inplace=True)
data.head()

Top1  ...                         Top25
0  A  hindrance to operations   extracts from the...  ...            Recovering a title
1                                          Scorecard  ...  Millennium bug fails to bite
2                  Coventry caught on counter by Flo  ...                  Useful links
3                      Pilgrim knows how to progress  ...   Lessons of law s hard heart
4                               Hitches and Horlocks  ...                    Gone aloft

[5 rows x 25 columns]

In [ ]:
data.columns

Index(['Top1', 'Top2', 'Top3', 'Top4', 'Top5', 'Top6', 'Top7', 'Top8', 'Top9',
       'Top10', 'Top11', 'Top12', 'Top13', 'Top14', 'Top15', 'Top16', 'Top17',
       'Top18', 'Top19', 'Top20', 'Top21', 'Top22', 'Top23', 'Top24', 'Top25'],
      dtype='object')

In [ ]:
#Lowering all the sentences

In [ ]:
for col in data.columns:
    data[col]=data[col].str.lower()
data.head()

Top1  ...                         Top25
0  a  hindrance to operations   extracts from the...  ...            recovering a title
1                                          scorecard  ...  millennium bug fails to bite
2                  coventry caught on counter by flo  ...                  useful links
3                      pilgrim knows how to progress  ...   lessons of law s hard heart
4                               hitches and horlocks  ...                    gone aloft

[5 rows x 25 columns]

In [ ]:
headlines = []
for row in range(0,len(data.index)):
    headlines.append(' '.join(str(x) for x in data.iloc[row,0:25]))

In [ ]:
headlines

['a  hindrance to operations   extracts from the leaked reports scorecard hughes  instant hit buoys blues jack gets his skates on at ice cold alex chaos as maracana builds up for united depleted leicester prevail as elliott spoils everton s party hungry spurs sense rich pickings gunners so wide of an easy target derby raise a glass to strupar s debut double southgate strikes  leeds pay the penalty hammers hand robson a youthful lesson saints party like it s      wear wolves have turned into lambs stump mike catches testy gough s taunt langer escapes to hit     flintoff injury piles on woe for england hunters threaten jospin with new battle of the somme kohl s successor drawn into scandal the difference between men and women sara denver  nurse turned solicitor diana s landmine crusade put tories in a panic yeltsin s resignation caught opposition flat footed russian roulette sold out recovering a title',
 'scorecard the best lake scene leader  german sleaze inquiry cheerio  boyo the main

In [ ]:
# Creating the Bag of Words model
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.ensemble import RandomForestClassifier

In [ ]:
## implement BAG OF WORDS
countvector=CountVectorizer(ngram_range=(2,3))
cv_train=countvector.fit_transform(headlines)

In [ ]:
# implement RandomForest Classifier
randomclassifier=RandomForestClassifier(n_estimators=200,criterion='entropy')
randomclassifier.fit(cv_train,train['Label'])

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='entropy', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=200,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [ ]:
## Transforming the Test Dataset
test_transform= []
for row in range(0,len(test.index)):
    test_transform.append(' '.join(str(x) for x in test.iloc[row,2:27]))
test_transform

['Most cases of cancer are the result of sheer bad luck rather than unhealthy lifestyles, diet or even inherited genes, new research suggests. Random mutations that occur in DNA when cells divide are responsible for two thirds of adult cancers across a wide range of tissues. Iran dismissed United States efforts to fight Islamic State as a ploy to advance U.S. policies in the region: "The reality is that the United States is not acting to eliminate Daesh. They are not even interested in weakening Daesh, they are only interested in managing it" Poll: One in 8 Germans would join anti-Muslim marches UK royal family\'s Prince Andrew named in US lawsuit over underage sex allegations Some 40 asylum-seekers refused to leave the bus when they arrived at their destination in rural northern Sweden, demanding that they be taken back to Malm or "some big city". Pakistani boat blows self up after India navy chase. All four people on board the vessel from near the Pakistani port city of Karachi are b

In [ ]:
cv_test = countvector.transform(test_transform)
predictions_cv = randomclassifier.predict(cv_test)

In [ ]:
## Import library to check accuracy for BAG OF WORDS
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

matrix_cv=confusion_matrix(test['Label'],predictions_cv)
print(matrix_cv)
score_cv=accuracy_score(test['Label'],predictions_cv)
print(score_cv)
report_cv=classification_report(test['Label'],predictions_cv)
print(report_cv)

[[135  51]
 [  1 191]]
0.8624338624338624
              precision    recall  f1-score   support

           0       0.99      0.73      0.84       186
           1       0.79      0.99      0.88       192

    accuracy                           0.86       378
   macro avg       0.89      0.86      0.86       378
weighted avg       0.89      0.86      0.86       378



86% accuracy for Bag of Words using Random Forest Classifier.

In [ ]:
# Creating the TF-IDF model
from sklearn.feature_extraction.text import TfidfVectorizer
#from sklearn.ensemble import RandomForestClassifier

In [ ]:
#implement TF-IDF
tfvector=TfidfVectorizer(ngram_range=(2,3))
tfd_train=tfvector.fit_transform(headlines)

In [ ]:
# implement RandomForest Classifier
randomclassifier=RandomForestClassifier(n_estimators=200,criterion='entropy')
randomclassifier.fit(tfd_train,train['Label'])

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='entropy', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=200,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [ ]:
tfd_test = tfvector.transform(test_transform)
predictions_tfd = randomclassifier.predict(tfd_test)

In [ ]:
## Import library to check accuracy for BAG OF WORDS
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

matrix_tfd=confusion_matrix(test['Label'],predictions_tfd)
print(matrix_tfd)
score_tfd=accuracy_score(test['Label'],predictions_tfd)
print(score_tfd)
report_tfd=classification_report(test['Label'],predictions_tfd)
print(report_tfd)

[[148  38]
 [ 23 169]]
0.8386243386243386
              precision    recall  f1-score   support

           0       0.87      0.80      0.83       186
           1       0.82      0.88      0.85       192

    accuracy                           0.84       378
   macro avg       0.84      0.84      0.84       378
weighted avg       0.84      0.84      0.84       378



84% accuracy for TF-IDF model using Random Forest CLassifier.

In [ ]:
# implement MultinomialBD for TFIDF
from sklearn.naive_bayes import MultinomialNB
nb=MultinomialNB()
nb.fit(tfd_train,train['Label'])

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [ ]:
tfd_test = tfvector.transform(test_transform)
predictions = nb.predict(tfd_test)

In [ ]:
predictions = nb.predict(tfd_test)
matrix=confusion_matrix(test['Label'],predictions)
print(matrix)
score=accuracy_score(test['Label'],predictions)
print(score)
report=classification_report(test['Label'],predictions)
print(report)

[[130  56]
 [  0 192]]
0.8518518518518519
              precision    recall  f1-score   support

           0       1.00      0.70      0.82       186
           1       0.77      1.00      0.87       192

    accuracy                           0.85       378
   macro avg       0.89      0.85      0.85       378
weighted avg       0.89      0.85      0.85       378



85% accuracy for TF-IDF model using Multinomial NB.